### PR TITLE
MAKE-521: ftdc2csv functions and additional tests

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: collector_test.go [hunk 1]
```diff
@@ -410,6 +415,63 @@ func TestFixedEncoding(t *testing.T) {
...
+		collector := NewWriterCollector(2, &noopWriter{})
+		_, err := collector.Write(nil)
+		assert.Error(t, err)
+	})
+	t.Run("RealDocument", func(t *testing.T) {
+		collector := NewWriterCollector(2, &noopWriter{})
+		doc, err := bson.NewDocument(bson.EC.Int64("one", 43), bson.EC.Int64("two", 5)).MarshalBSON()
+		require.NoError(t, err)
+		_, err = collector.Write(doc)
+		assert.NoError(t, err)
```
Comment (julianedwards) [inline collector_test.go]: Shouldn't you close the collector?

File: collector_test.go [hunk 2]
```diff
@@ -410,6 +415,63 @@ func TestFixedEncoding(t *testing.T) {
...
+			assert.Error(t, collector.Add(bson.NewDocument(bson.EC.Int64("one", 43), bson.EC.Int64("two", 5))))
 		})
 	}
 }
+
+func TestWriter(t *testing.T) {
+	t.Run("NilDocuments", func(t *testing.T) {
+		collector := NewWriterCollector(2, &noopWriter{})
+		_, err := collector.Write(nil)
+		assert.Error(t, err)
```
Comment (julianedwards) [inline collector_test.go]: Shouldn't you close the collector?

File: collector_streaming.go [hunk 3]
```diff
@@ -55,6 +55,9 @@ func (c *streamingCollector) Add(d *bson.Document) error {
 // particularly useful in the context of streaming collectors, which
 // flush data periodically and may have cached data.
 func FlushCollector(c Collector, writer io.Writer) error {
+	if writer == nil {
+		return errors.New("ivalid writer")
```
Comment (julianedwards) [inline collector_streaming.go]: invalid*

File: writer.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+type writerCollector struct {
+	writer    io.WriteCloser
+	collector *streamingDynamicCollector
+}
+
+func NewWriterCollector(chunkSize int, writer io.WriteCloser) io.WriteCloser {
+	return &writerCollector{
+		writer: writer,
+		collector: &streamingDynamicCollector{
```
Comment (julianedwards) [inline writer.go]: Why not use `NewStreamingDynamicCollector` here? https://github.com/mongodb/ftdc/blob/82de4a82e6d62196f8f1fed6c883fd738b0fada0/collector_streaming.go#L91

File: ftdc.go [hunk 5]
```diff
@@ -94,7 +94,7 @@ type Metric struct {
 	// Values is an array of each value collected for this metric.
 	// During decoding, this attribute stores delta-encoded
 	// values, but those are expanded during decoding and should
-	// never be visible to user.
+	// never be visible toser.
```
Comment (julianedwards) [inline ftdc.go]: to user*

File: csv.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	"io"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/mongodb/mongo-go-driver/bson"
+	"github.com/pkg/errors"
+)
+
+func (c *Chunk) getFieldNames() []string {
```
Comment (julianedwards) [inline csv.go]: I would argue to put this function and `getRecord` in ftdc.go where Chunk is defined.

File: csv.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	return writer, nil
+}
+
+// DumpCSV writes a sequence of chunks to CSV files, creating new
+// files if the iterator detects a schema change, using only the
+// number of fields in the chunk to detect schema changes. DumpCSV
+// writes a header row to each file.
+//
+// The file names are constructed as "prefix.<count>.csv".
+func DumpCSV(ctx context.Context, iter *ChunkIterator, prefix string) error {
```
Comment (julianedwards) [inline csv.go]: maybe call this `DynamicCSV` to be consistent with `DynamicCollector`?

File: collector_test.go (hunk 2 above)
Comment (tychoish) [reply collector_test.go]: done

File: writer.go (hunk 4 above)
Comment (tychoish) [reply writer.go]: becasue it returns a Collector type and I wanted it to be of the concrete type.

File: csv.go (hunk 7 above)
Comment (tychoish) [reply csv.go]: I kind of like it as is. Its less about dynamic/not and more about the chunking.

maybe it makes sense to add a max number of rows per file as well to make it less about schema change.

File: writer.go (hunk 4 above)
Comment (julianedwards) [reply writer.go]: ack

File: collector_test.go (hunk 1 above)
Comment (tychoish) [reply collector_test.go]: done

File: csv.go (hunk 6 above)
Comment (tychoish) [reply csv.go]: the methods on the chunk type are pretty distributed closer to call site. I think I agree with you in the big picture about this kind of change, but I'd rather leave it the way it is.

File: csv.go (hunk 7 above)
Comment (julianedwards) [reply csv.go]: Well it only creates a new csv file when the schema changes, so as is it is all about schema change. 

If we do add max number of rows, we would still have to change when the schema changes, correct?

File: csv.go (hunk 6 above)
Comment (julianedwards) [reply csv.go]: sounds good

File: csv.go (hunk 7 above)
Comment (tychoish) [reply csv.go]: yep.

**Pull request merged**